packages/gatsby/cache-dir/develop-static-entry.js

```diff
@@ -79,6 +91,109 @@ export default (pagePath, callback) => {
     postBodyComponents = components
   }
 
+  const getPageDataPath = path => {
+    const fixedPagePath = path === `/` ? `index` : path
+    return join(`page-data`, fixedPagePath, `page-data.json`)
+  }
+
+  const getPageData = pagePath => {
+    const pageDataPath = getPageDataPath(pagePath)
+    const absolutePageDataPath = join(process.cwd(), `public`, pageDataPath)
+    const pageDataRaw = fs.readFileSync(absolutePageDataPath)
+
+    try {
+      return JSON.parse(pageDataRaw.toString())
+    } catch (err) {
+      return null
+    }
+  }
+
+  const pageData = getPageData(pagePath)
+
+  const {
+    componentChunkName,
+    //  staticQueryHashes = []
+  } = pageData
+
+  const createElement = React.createElement
+
+  class RouteHandler extends React.Component {
+    render() {
+      const props = {
+        ...this.props,
+        ...pageData.result,
+        params: {
+          ...grabMatchParams(this.props.location.pathname),
+          ...(pageData.result?.pageContext?.__params || {}),
+        },
+        // pathContext was deprecated in v2. Renamed to pageContext
+        pathContext: pageData.result ? pageData.result.pageContext : undefined,
+      }
+
+      const pageElement = createElement(
+        syncRequires.components[componentChunkName],
+        props
+      )
+
+      const wrappedPage = apiRunner(
+        `wrapPageElement`,
+        { element: pageElement, props },
+        pageElement,
+        ({ result }) => {
+          return { element: result, props }
+        }
+      ).pop()
+
+      return wrappedPage
+    }
+  }
+
+  const routerElement = (
+    <ServerLocation url={`${__BASE_PATH__}${pagePath}`}>
+      <Router id="gatsby-focus-wrapper" baseuri={__BASE_PATH__}>
+        <RouteHandler path="/*" />
+      </Router>
+      <div {...RouteAnnouncerProps} />
+    </ServerLocation>
+  )
+
+  const bodyComponent = apiRunner(
+    `wrapRootElement`,
+    { element: routerElement, pathname: pagePath },
+    routerElement,
+    ({ result }) => {
+      return { element: result, pathname: pagePath }
+    }
+  ).pop()
+
+  if (process.env.GATSBY_EXPERIMENTAL_DEV_SSR) {
+    // Let the site or plugin render the page component.
+    apiRunner(`replaceRenderer`, {
+      bodyComponent,
+      replaceBodyHTMLString,
+      setHeadComponents,
+      setHtmlAttributes,
+      setBodyAttributes,
+      setPreBodyComponents,
+      setPostBodyComponents,
+      setBodyProps,
+      pathname: pagePath,
+      pathPrefix: __PATH_PREFIX__,
+    })
+  }
+
+  // If no one stepped up, we'll handle it.
+  if (!bodyHtml) {
+    try {
+      if (process.env.GATSBY_EXPERIMENTAL_DEV_SSR) {
+        bodyHtml = renderToString(bodyComponent)
+      }
+    } catch (e) {
+      // ignore @reach/router redirect errors
+      if (!isRedirect(e)) throw e
```

Comment:
No `reporter.panic` structured error?

Reply:
Throwing is what we want as we catch SSR errors in https://github.com/gatsbyjs/gatsby/pull/27432/files#diff-1635212f76bd1e3d30473f797fbbc58ef27a64bdda0af0870d49ae7476de03f9R171 to show the user a nice dev ssr error page in the browser.